File: notes_and_examples/file_search.java
import java.io.IOException;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.ArrayList;
import java.util.List;

public class FileSearch {
    public static void main(String[] args) {
        String directoryToSearch = "path/to/directory"; // Replace with the directory you want to search
        String fileExtensionToFind = ".txt"; // Replace with the file extension you want to search for

        List<String> foundFiles = searchFiles(directoryToSearch, fileExtensionToFind);

        if (foundFiles.isEmpty()) {
            System.out.println("No files found with the extension '" + fileExtensionToFind + "' in the directory.");
        } else {
            System.out.println("Found " + foundFiles.size() + " files with the extension '" + fileExtensionToFind + "':");
            for (String filePath : foundFiles) {
                System.out.println(filePath);
            }
        }
    }

    public static List<String> searchFiles(String directory, String fileExtension) {
        List<String> foundFiles = new ArrayList<>();
        Path startingDir = Paths.get(directory);
        try {
            Files.walkFileTree(startingDir, new SimpleFileVisitor<Path>() {
                @Override
                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
                    if (file.toString().endsWith(fileExtension)) {
                        foundFiles.add(file.toString());
                    }
                    return FileVisitResult.CONTINUE;
                }

                @Override
                public FileVisitResult visitFileFailed(Path file, IOException exc) {
                    return FileVisitResult.CONTINUE;
                }
            });
        } catch (IOException e) {
            e.printStackTrace();
        }
        return foundFiles;
    }
}

File: notes_and_examples/file_search.py
# Python program that performs a file search operation within a specified directory and its subdirectories using the os module

import os

def search_files(directory, file_extension):
    found_files = []
    for dirpath, dirnames, filenames in os.walk(directory):
        for filename in filenames:
            if filename.endswith(file_extension):
                found_files.append(os.path.join(dirpath, filename))
    return found_files

# Set the directory to search and the file extension to look for
directory_to_search = '/path/to/directory'  # Replace with the directory you want to search
file_extension_to_find = '.txt'  # Replace with the file extension you want to search for

# Search for files with the specified extension in the given directory
result = search_files(directory_to_search, file_extension_to_find)

if result:
    print(f"Found {len(result)} files with the extension '{file_extension_to_find}':")
    for file_path in result:
        print(file_path)
else:
    print(f"No files found with the extension '{file_extension_to_find}' in the directory.")

"""

This script uses the os.walk() function to traverse through the directory and its subdirectories, 
searching for files with the specified file extension. 
It then displays the paths of the found files.

"""
